src/python/Utils/Throttled.py

```diff
@@ -16,41 +36,130 @@
         ctr = self.throttle._decUser(self.user)
-        if self.debug:
-            msg = "Exiting throttled function with counter %d for user %s" \
-                  % (ctr, self.user)
-            self.throttle.logger.debug(msg)
+        msg = "Exiting throttled function with counter %d for user %s" \
+                % (ctr, self.user)
+        self.throttle.logger.debug(msg)
+
+class _ThrottleTimeCounter(object):
+    """
+    _ThrottleTimeCounter class defines time range throttled mechanism
+    """
+
+    def __init__(self, throttle, user, trange):
+        self.throttle = throttle
+        self.user = user
+        self.trange = trange
+
+    def __enter__(self):
+        "Define enter method for `with` context"
+        ctr = self.throttle._incUser(self.user, self.trange)
+        if ctr > self.throttle.getLimit():
+            self.throttle._decUser(self.user)
+            msg = "The current number of active operations for this resource"
+            msg += " exceeds the limit of %d for user %s in last %s sec" \
+                   % (self.throttle.getLimit(), self.user, self.trange)
+            raise cherrypy.HTTPError("429 Too Many Requests", msg)
+
+    def __exit__(self, exc_type, exc_value, exc_traceback):
+        "Define exit method for `with` context"
+        pass
+
+
+class UserThrottleTime(object):
+    """
+    UserThrottle class defines how to handle throttle time range based mechanism.
+    """
 
+    def __init__(self, limit=3):
+        self.lock = threading.Lock()
+        self.users = {}
+        self.users_time = {}
+        self.limit = limit
+
+    def getLimit(self):
+        "Return throttle limit"
+        return self.limit
+
+    def throttleContext(self, user, trange=60):
+        """
+        Define throttle context via _ThrottleTimeCounter class
+        which will count number of requests given user made in trange
+        :param user: user name
+        :param trange: time range while keep user access history
+        """
+        return _ThrottleTimeCounter(self, user, trange)
+
+    def make_throttled(self, trange=60):
+        "decorator for throttled context"
+        def throttled_decorator(fn):
+            def throttled_wrapped_function(*args, **kw):
+                username = cherrypy.request.user.get('login', 'Unknown') \
+                        if hasattr(cherrypy.request, 'user') else 'Unknown'
+                with self.throttleContext(username, trange):
+                    return fn(*args, **kw)
+            return throttled_wrapped_function
+        return throttled_decorator
+
+    def reset(self, user):
+        "Reset user activities, i.e. access counter and last time of access"
+        self.users[user] = 0
+        self.users_time[user] = time.time()
+
+    def _incUser(self, user, trange=60):
+        "increment user count"
+        with self.lock:
+            self.users.setdefault(user, 0)
+            self.users_time.setdefault(user, time.time())
```

Comment:
Just a small python suggestion as `setdefault` returns the resulting value in the `dict`

```suggestion
            last_time = self.users_time.setdefault(user, time.time())
```

Reply:
done